enterprise/internal/campaigns/webhooks_test.go

```diff
@@ -193,6 +193,192 @@ func testGitHubWebhook(db *sql.DB) func(*testing.T) {
+		}
+
+		bitbucketSource, err := repos.NewBitbucketServerSource(githubExtSvc, cf, nil)
+		if err != nil {
+			t.Fatal(t)
+		}
+
+		repoChan := make(chan repos.SourceResult)
+		go func() {
+			bitbucketSource.ListRepos(ctx, repoChan)
+			close(repoChan)
+		}()
+
+		var bitbucketRepo *repos.Repo
+		for result := range repoChan {
+			if result.Err != nil {
+				t.Fatal(result.Err)
+			}
+			if result.Repo == nil {
+				continue
+			}
+			if result.Repo.Name == "bitbucket.sgdev.org/SOUR/automation-testing" {
+				bitbucketRepo = result.Repo
+			}
+		}
+
+		if bitbucketRepo == nil {
+			t.Fatal("repo not found")
+		}
+
+		err = repoStore.UpsertRepos(ctx, bitbucketRepo)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		store := NewStoreWithClock(db, clock)
+
+		campaign := &campaigns.Campaign{
+			Name:            "Test campaign",
+			Description:     "Testing THE WEBHOOKS",
+			AuthorID:        userID,
+			NamespaceUserID: userID,
+		}
+
+		err = store.CreateCampaign(ctx, campaign)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		changesets := []*campaigns.Changeset{
+			{
+				RepoID:              bitbucketRepo.ID,
+				ExternalID:          "69",
+				ExternalServiceType: bitbucketRepo.ExternalRepo.ServiceType,
+				CampaignIDs:         []int64{campaign.ID},
+			},
+			{
+				RepoID:              bitbucketRepo.ID,
+				ExternalID:          "19",
+				ExternalServiceType: bitbucketRepo.ExternalRepo.ServiceType,
+				CampaignIDs:         []int64{campaign.ID},
+			},
+		}
+
+		err = store.CreateChangesets(ctx, changesets...)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		err = SyncChangesets(ctx, repoStore, store, cf, changesets...)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		hook := NewBitbucketServerWebhook(store, repoStore, clock, "testhook")
+
+		fixtureFiles, err := filepath.Glob("testdata/fixtures/webhooks/bitbucketserver/*.json")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		for _, fixtureFile := range fixtureFiles {
+			_, name := path.Split(fixtureFile)
+			name = strings.TrimSuffix(name, ".json")
+			t.Run(name, func(t *testing.T) {
+				_, err = db.Exec("ALTER SEQUENCE changeset_events_id_seq RESTART")
+				if err != nil {
+					t.Fatal(err)
+				}
+				_, err = db.Exec("TRUNCATE TABLE changeset_events")
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				tc := loadWebhookTestCase(t, fixtureFile)
+
+				// Send all events twice to ensure we are idempotent
+				for i := 0; i < 2; i++ {
+					for _, event := range tc.Payloads {
+						req, err := http.NewRequest("POST", "", bytes.NewReader(event.Data))
+						if err != nil {
+							t.Fatal(err)
+						}
+						req.Header.Set("X-Hub-Signature", sign(t, event.Data, []byte(secret)))
+
+						rec := httptest.NewRecorder()
+						hook.ServeHTTP(rec, req)
+						resp := rec.Result()
+
+						if resp.StatusCode != http.StatusOK {
+							t.Fatalf("Non 200 code: %v", resp.StatusCode)
+						}
+					}
+				}
+
+				have, _, err := store.ListChangesetEvents(ctx, ListChangesetEventsOpts{Limit: 1000})
```

Comment:
I know, I know, but... let's just be safe 😎 

```suggestion
				have, _, err := store.ListChangesetEvents(ctx, ListChangesetEventsOpts{Limit: -1})
```